src/util/file_reader.js

```diff
@@ -44,29 +63,66 @@ class NewlineReader {
     async nextline() {
         if (!this.fh) await this.init();
 
-        if (this.readresults.length) return this.readresults.shift();
-        if (this.eof) return null;
+        while (!this.eof) {
+            // extract next line if terminated in current buffer
+            if (this.start < this.end) {
+                const term_idx = this.buf.subarray(this.start, this.end).indexOf(10);
+                if (term_idx >= 0) {
+                    if (this.overflow_state) {
+                        console.warn('line too long finally terminated:', this.info());
+                        this.overflow_state = false;
+                        this.start += term_idx + 1;
+                        continue;
+                    }
+
+                    const line = this.buf.toString('utf8', this.start, this.start + term_idx);
+                    this.start += term_idx + 1;
+                    return line;
+                }
+            }
 
-        // Will keep reading till we find at least one new line character
-        while (!this._partialread.includes('\n')) {
-            const read = await this.fh.read(this.fs_context, this.buf, 0, this.buf.length, this.readoffset);
-            if (read === 0) {
-                this.eof = true;
-                return null;
+            // relocate existing data to offset 0 in buf
+            if (this.start > 0) {
+                const n = this.buf.copy(this.buf, 0, this.start, this.end);
+                this.start = 0;
+                this.end = n;
             }
 
-            this.readoffset += read;
+            // check limits
+            if (this.buf.length <= this.end) {
+                if (!this.skip_overflow_lines) {
+                    throw new Error("line too long or non terminated");
+                }
 
-            this._partialread += this.buf.subarray(0, read).toString('utf-8');
-        }
+                console.warn('line too long or non terminated:', this.info());
+                this.end = 0;
+                this.start = 0;
+                this.overflow_state = true;
+            }
+
+            // read from file
+            const avail = this.buf.length - this.end;
+            const read = await this.fh.read(this.fs_context, this.buf, this.end, avail, this.readoffset);
+            if (!read) {
+                this.eof = true;
 
-        // readresults will contain >= 1 result or else we would have kept looping above
-        this.readresults = this._partialread.split('\n').slice(0, -1);
+                // what to do with the leftover in the buffer on eof
+                if (this.end > this.start) {
+                    if (this.skip_leftover_line) {
+                        console.warn("leftover at eof:", this.info());
+                    } else {
```

Comment:
we need to check also the overflow state here and not return it as a line
```suggestion
                    } else if (this.overflow_state) {
                        console.warn('line too long finally terminated at eof:', this.info());
                    } else {
```